### ISSUE TITLE
[23] Formatter support for markdown doc comments

### ISSUE THREAD

**Issue by stephan-herrmann** [eclipse-jdt/eclipse.jdt.core#2851]
To complete #2845  we'd need to upgrade the formatter, too.

Currently, trying to format a document with markdown comments will throw exceptions like:
```
java.lang.AssertionError
	at org.eclipse.jdt.internal.formatter.TokenManager.firstIndexIn(TokenManager.java:194)
	at org.eclipse.jdt.internal.formatter.CommentsPreparator.visit(CommentsPreparator.java:590)
	at org.eclipse.jdt.core.dom.Javadoc.accept0(Javadoc.java:191)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3313)
	at org.eclipse.jdt.internal.formatter.DefaultCodeFormatter.prepareComments(DefaultCodeFormatter.java:431)
	at org.eclipse.jdt.internal.formatter.DefaultCodeFormatter.prepareFormattedCode(DefaultCodeFormatter.java:233)
```

or
```
java.lang.IndexOutOfBoundsException: Index 29 out of bounds for length 29
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:365)
	at java.base/java.util.ArrayList.get(ArrayList.java:428)
	at org.eclipse.jdt.internal.formatter.TokenManager.get(TokenManager.java:79)
	at org.eclipse.jdt.internal.formatter.TokenManager.findIndex(TokenManager.java:173)
	at org.eclipse.jdt.internal.formatter.TokenManager.firstIndexIn(TokenManager.java:193)
	at org.eclipse.jdt.internal.formatter.CommentsPreparator.visit(CommentsPreparator.java:590)
	at org.eclipse.jdt.core.dom.Javadoc.accept0(Javadoc.java:191)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3313)
	at org.eclipse.jdt.internal.formatter.DefaultCodeFormatter.prepareComments(DefaultCodeFormatter.java:431)
	at org.eclipse.jdt.internal.formatter.DefaultCodeFormatter.prepareFormattedCode(DefaultCodeFormatter.java:233)
```
etc.

**Comment by stephan-herrmann** [comment]
@mateusz-matela   I haven't "seen" you much lately :) -- assumeably because the formatter just works :) :) 

Could you help us in this issue? It's quite possible, that parsers create inconsistent positions or such, which disturb the formatter. It would already help a great deal, if you point out which assumptions in the formatter are violated by recent parser changes.

**Comment by stephan-herrmann** [comment]
The first error (AssertionError) is caused in this section of CommentsPreparator:
```
	public boolean visit(Javadoc node) {
		this.noFormatOpenTagStartIndex = -1;
		this.formatCodeOpenTagEndIndex = -1;
		this.lastFormatCodeClosingTagIndex = -1;
		this.commonAttributeAnnotations.clear();
		this.ctm = null;

		int commentIndex = this.tm.firstIndexIn(node, TokenNameCOMMENT_JAVADOC);
```
Here we should select `TokenNameCOMMENT_MARKDOWN` if the `Javadoc` is of type markdown.

@jarthana This seems to be the trigger for needing that flag in DOM indeed :)

I might give it a quick attempt, to see how far it will take us towards formatting.

**Comment by stephan-herrmann** [comment]
In the interest of time I merged the minimal fix #2858  (in 3 separate commits, to separate DOM changes from formatter changes).

Further improvements can be handled by separate issues.
